KEP-2853: Initial Draft of branch rename for k/k (#3053)
kubernetes/enhancements: keps/sig-release/2853-kubernetes-kubernetes-repository-branch-rename/README.md

Changing the branch name might cause issues for downstream consumers, our contributors, and the community in general.

For that, we are creating this KEP to have a plan to address the issues and communication.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->
The main goal is to be more inclusive using better naming.

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->
This applies only to the main Kubernetes GitHub repository [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes).
Other repositories will follow a similar approach but at their timeframe.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->
- We aim to make the change during the start of v1.25 release (spring 2022).

> - Perform a Survey to gather information on downstream consumers and how this might affects their workflow. See [Survey Questions](#survey-questions).

Reviewer: I'm wondering should the survey cover the upstream consumers as well. For example, there might be SIGs and/or projects under `kubernetes*` orgs that are affected by this change. IMO, we also want to be sure that they can migrate by the start of the v1.25 release cycle. We should also note that the upstream consumers might be affected by the enhancements and code freeze.

- Announce the change in the kubernetes-dev and kubernetes-announcement mailing list and in a blog post. As well as use Twitter and other social media to announce.
- Change all open Pull Requests to `Draft mode` (so that when we rename the branch, tests will not be triggered, and we avoid a massive queue in the CI infrastructure).
- Update all Prow jobs that use `kubernetes/kubernetes` and use the `master` branch to listen to the `main` branch as well
- Follow the guide on https://github.com/kubernetes/community/blob/master/github-management/default-branch-migration.md to rename the repository branch name.
- After the branch rename, announce that in the mailing list
- Update all Prow jobs that use `kubernetes/kubernetes` and to remove the `master` branch
- Convert on demand the open Pull Requests from the `Draft mode` to `Review ready mode`

References:
- https://github.com/kubernetes/org/issues/2222
- https://github.com/kubernetes/enhancements/issues/2853
- https://github.com/kubernetes/kubernetes/issues/105601
- https://github.com/kubernetes/community/blob/master/github-management/default-branch-migration.md

- https://github.com/github/renaming

> ### Risks and Mitigations

Reviewer: Have we done any research on how is it going to affect us (speaking of Release Engineering)? For example:

* What changes we might need to do to krel and the release tooling?
* What about the release notes tool?
* What about the version markers (could they be affected by this)?

Author: > What changes we might need to do to krel and the release tooling?
> What about the release notes tool?

we will need to update some default const we have to use main, or do some fallback if fail, like try master and if fail go to main, or otherwise

> What about the version markers (could they be affected by this)?

regarding this maybe @saschagrunert or @justaugustus can give some opinion, I'm not expert in this topic :(

Reviewer 2: Any tooling would require to be adapted. So there is a risk we miss some of them, which can be mitigated by grepping with multiple eyes :)

Author: will add this as a risk and mitigation plans

Changing the branch name might cause issues for downstream consumers, our contributors, and the community in general.

For that, we are creating this KEP to have a plan to address the issues and communication.

### Goals

<!--
List the specific goals of the KEP. What is it trying to achieve? How will we
know that this has succeeded?
-->
The main goal is to be more inclusive using better naming.

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->
This applies only to the main Kubernetes GitHub repository [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes).
Other repositories will follow a similar approach but at their timeframe.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->

> - The change can be applied in the middle of December 2021. This is a good time because we will have less activity due to the Christmas and new year celebrations

Reviewer 2: I'd propose to change it when everything is back from christmas, in the middle of a release cycle or directly at the beginning. Let's aim for merging the KEP before mid-december, this gives us enough room for communicating the change afterwards.

How about choosing the start of the v1.25 (spring 2022) release cycle for the branch rename? For example, directly after the `master` branch opens on `Thaw`?

Author: sounds a good idea

Author: updated

> - Announce the change in the kubernetes-dev and kubernetes-announcement mailing list and maybe in a blog post. As well as use Twitter and other social media to announce.

Reviewer 2: ```suggestion
- Announce the change in the kubernetes-dev and kubernetes-announcement mailing list and in a blog post. As well as use Twitter and other social media to announce.
```

> - Change all open Pull Requests to the `Draft mode` with that when we rename the branch, the tests will not be triggered and will avoid a massive queue in the CI infrastructure.

Reviewer 2: Have we verified how the branch renaming behaves on existing PRs? I would expect that GitHub just links the new target branch, so why should it re-trigger the CI?

Author: when you change the branch, looks like it get triggered, I saw this in the other renames I did.

- Update all Prow jobs that use `kubernetes/kubernetes` and use the `master` branch to listen to the `main` branch as well
- Follow the guide on https://github.com/kubernetes/community/blob/master/github-management/default-branch-migration.md to rename the repository branch name.
- After the branch rename, announce that in the mailing list
- Update all Prow jobs that use `kubernetes/kubernetes` and to remove the `master` branch
- Convert on demand the open Pull Requests from the `Draft mode` to `Review ready mode`

References:
- https://github.com/kubernetes/org/issues/2222
- https://github.com/kubernetes/enhancements/issues/2853
- https://github.com/kubernetes/kubernetes/issues/105601
- https://github.com/kubernetes/community/blob/master/github-management/default-branch-migration.md

- https://github.com/github/renaming

> 

Reviewer 2: We should additionally gather input from downstream consumers how it affects their workflow. Let's do a survey for this and add the questions as part of this KEP. For example:

- On which companies behalf do you submit the response?
- Would a k/k branch rename affect your downstream workflow? If yes, how?
- How much time would you need in advance before the migration happens?

Author: added the initial questions

marked as `provisional` is a working document and subject to change. You can
denote sections that are under active debate as follows:

```
<<[UNRESOLVED optional short context or usernames ]>>
Stuff that is being argued.
<<[/UNRESOLVED]>>
```

When editing KEPS, aim for tightly-scoped, single-topic PRs to keep discussions
focused. If you disagree with what is already in a document, open a new PR
with suggested changes.

One KEP corresponds to one "feature" or "enhancement" for its whole lifecycle.
You do not need a new KEP to move from beta to GA, for example. If
new details emerge that belong in the KEP, edit the KEP. Once a feature has become
"implemented", major changes should get new KEPs.

The canonical place for the latest set of instructions (and the likely source
of this file) is [here](/keps/NNNN-kep-template/README.md).

**Note:** Any PRs to move a KEP to `implementable`, or significant changes once
it is marked `implementable`, must be approved by each of the KEP approvers.
If none of those approvers are still appropriate, then changes to that list
should be approved by the remaining approvers and/or the owning SIG (or
SIG Architecture for cross-cutting KEPs).
-->

> # KEP-NNNN: Your short, descriptive title

Reviewer 3: 2853, right?

## Summary

<!--
This section is incredibly important for producing high-quality, user-focused
documentation such as release notes or a development roadmap. It should be
possible to collect this information before implementation begins, in order to
avoid requiring implementors to split their attention between writing release
notes and implementing the feature itself. KEP editors and SIG Docs
should help to ensure that the tone and content of the `Summary` section is
useful for a wide audience.

A good summary is probably at least a paragraph in length.

Both in this section and below, follow the guidelines of the [documentation
style guide]. In particular, wrap lines to a reasonable length, to make it
easier for reviewers to cite specific portions, and to minimize diff churn on
updates.

[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
-->

Many communities, both on GitHub and in the wider Git community, are considering renaming the default branch name of their repository from `master`. The Kubernetes GitHub repositories is gradually renaming the default branch of our own repositories from `master` to `main`.

> This KEP is to we rename the branch for the main repository in the Kubernetes community.

Reviewer 3: ```suggestion
This KEP proposes that we rename the branch for the principal code repository in the Kubernetes project.
```

Author: thanks so much for your review Tim!

know that this has succeeded?
-->
The main goal is to be more inclusive using better naming.

### Non-Goals

<!--
What is out of scope for this KEP? Listing non-goals helps to focus discussion
and make progress.
-->
This applies only to the main Kubernetes GitHub repository [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes).
Other repositories will follow a similar approach but at their timeframe.

## Proposal

<!--
This is where we get down to the specifics of what the proposal actually is.
This should have enough detail that reviewers can understand exactly what
you're proposing, but should not include things like API designs or
implementation. What is the desired outcome and how do we measure success?.
The "Design Details" section below is for the real
nitty-gritty.
-->
- We aim to make the change during the start of v1.25 release (spring 2022).
- Perform a Survey to gather information on downstream consumers and how this might affects their workflow. See [Survey Questions](#survey-questions).
- Announce the change in the kubernetes-dev and kubernetes-announcement mailing list and in a blog post. As well as use Twitter and other social media to announce.

> - Change all open Pull Requests to the `Draft mode` with that when we rename the branch, the tests will not be triggered and will avoid a massive queue in the CI infrastructure.

Reviewer 3: ```suggestion
- Change all open Pull Requests to `Draft mode` (so that when we rename the branch, tests will not be triggered, and we avoid a massive queue in the CI infrastructure).
```

- Update all Prow jobs that use `kubernetes/kubernetes` and use the `master` branch to listen to the `main` branch as well
- Follow the guide on https://github.com/kubernetes/community/blob/master/github-management/default-branch-migration.md to rename the repository branch name.
- After the branch rename, announce that in the mailing list
- Update all Prow jobs that use `kubernetes/kubernetes` and to remove the `master` branch
- Convert on demand the open Pull Requests from the `Draft mode` to `Review ready mode`

References:
- https://github.com/kubernetes/org/issues/2222
- https://github.com/kubernetes/enhancements/issues/2853
- https://github.com/kubernetes/kubernetes/issues/105601
- https://github.com/kubernetes/community/blob/master/github-management/default-branch-migration.md

- https://github.com/github/renaming

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.

> -->

Reviewer 3: As a contributor, I have to remember to use `master` when working with repositories that
haven't yet switched.
Switching k/k to `main` will reduce the burden slightly.

you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.
-->

- [ ] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name:
  - Components depending on the feature gate:
- [ ] Other
  - Describe the mechanism:
  - Will enabling / disabling the feature require downtime of the control
    plane?
  - Will enabling / disabling the feature require downtime or reprovisioning
    of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).

###### Does enabling the feature change any default behavior?

<!--
Any change of default behavior may be surprising to users or break existing
automations, so be extremely careful here.
-->

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

<!--
Describe the consequences on existing workloads (e.g., if this is a runtime
feature, can it break the existing applications?).

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.

> -->

Reviewer 3: We should answer this, I think.

- Follow the guide on https://github.com/kubernetes/community/blob/master/github-management/default-branch-migration.md to rename the repository branch name.
- After the branch rename, announce that in the mailing list
- Update all Prow jobs that use `kubernetes/kubernetes` and to remove the `master` branch
- Convert on demand the open Pull Requests from the `Draft mode` to `Review ready mode`

References:
- https://github.com/kubernetes/org/issues/2222
- https://github.com/kubernetes/enhancements/issues/2853
- https://github.com/kubernetes/kubernetes/issues/105601
- https://github.com/kubernetes/community/blob/master/github-management/default-branch-migration.md

- https://github.com/github/renaming

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->
As a contributor, I have to remember to use `master` when working with repositories that
haven't yet switched.
Switching k/k to `main` will reduce the burden slightly.

#### Story 1

> #### Story 2

Reviewer 2: Let's remove those or separate the stories under their headlines.

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

> 

Reviewer 2: Add a risk that we may break downstream consumers, we use the survey as mitigation as well as the schedule to let them prepare their build systems for the change.
